source/extensions/tracers/opentelemetry/span_context_extractor.cc

```diff
@@ -87,7 +91,20 @@ absl::StatusOr<SpanContext> SpanContextExtractor::extractSpanContext() {
   // See https://w3c.github.io/trace-context/#trace-flags.
   char decoded_trace_flags = absl::HexStringToBytes(trace_flags).front();
   bool sampled = (decoded_trace_flags & 1);
-  SpanContext span_context(version, trace_id, parent_id, sampled);
+
+  // If a tracestate header is received without an accompanying traceparent header,
+  // it is invalid and MUST be discarded.
+  // See https://www.w3.org/TR/trace-context/#processing-model-for-working-with-trace-context
+  absl::string_view tracestate = "";
+  // TODO: what if tracestate is in multiple headers? Spec says we should combine, but Envoy
+  // currently only surfaces the first seen.
```

Comment:
See `get`, which returns a `GetResult` with all header values.

Reply:
Thanks for that pointer! Did a little digging, and it looks like `ConnectionManagerImpl::ActiveStream::traceRequest` calls `startSpan` with its `RequestHeaderMapPtr request_headers_` (the `RequestHeaderMap` containing the `get`).  `Driver::startSpan` is defined with just `Tracing::TraceContext` though, which only has `getByKey` and `forEach`.

With that in mind, I added some handling for this (and a test) with iteration over all the headers. Doesn't seem very efficient to me, but I think this works for the multiple headers case. Happy to improve if you have any suggestions 👍

Reply:
Maybe fix `TraceContext` to deal with the appropriate header-specific aggregation behavior. I think any tracer interacting with tracestate will need this behavior. I think in most places that we see header get-by-key and the value is singular and not aggregated, there is some code smell that we haven't fully thought through the valid wire combinations.